### PR TITLE
Propose some functional helpers

### PR DESCRIPTION
**Pull request by inlined**
While reviewing #3690 I noticed that (A) one of the functions are generic enough to go in a library and (B) I really wish we had tools like zip/flatten/etc here. While writing my own code, I wished TypeScript could either assert or detect that code was dead due to type guards. Now we have all the above!

I leaned into using generator functions because they work just as well in `for-of` loops with less syntactic goop. LMK if you feel we should be returning arrays instead.

Feel free to tell me that these should be named differently, that we're missing something obvious, or that if God wanted us to have better list comprehensions they would have joined TC39.

### CODE REVIEW

**Review by bkendall — APPROVED** [review]
LGTM. First one to venture into making generators, but I have no problem with that 👍

File: src/functional.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+/** Flattens an object or array. */
+export function flatten<T>(
+  objOrArr: Record<string, unknown> | unknown[]
+): Generator<[string, unknown]> | Generator<T> {
+  if (Array.isArray(objOrArr)) {
+    return flattenArray<T>(objOrArr);
+  } else {
+    return flattenObject(objOrArr);
+  }
+}
```
Comment (taeold) [inline src/functional.ts:53]: yessss

File: src/test/functional.spec.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+describe("functional", () => {
+  describe("flatten", () => {
+    it("Can iterate an empty object", () => {
+      expect([...f.flatten({})]).to.deep.equal([]);
+    });
+
+    it("Can iterate an object that's already flat", () => {
+      expect([...f.flatten({ a: "b" })]).to.deep.equal([["a", "b"]]);
+    });
```
Comment (taeold) [inline src/test/functional.spec.ts:15]:
```suggestion
    it("can iterate an empty object", () => {
      expect([...f.flatten({})]).to.deep.equal([]);
    });

    it("can iterate an object that's already flat", () => {
      expect([...f.flatten({ a: "b" })]).to.deep.equal([["a", "b"]]);
    });
```

**Review by taeold — APPROVED** [review]
👏 This is so much fun.

**Pull request merged**